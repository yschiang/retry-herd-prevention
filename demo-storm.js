// demo-storm.js - Shows what happens WITHOUT throttling (RETRY STORM!)
// This demonstrates the problem we're solving

import { setTimeout as sleep } from 'node:timers/promises';

console.log('='.repeat(60));
console.log('‚ö†Ô∏è  RETRY STORM DEMO (WITHOUT PROTECTION)');
console.log('='.repeat(60));
console.log('\nScenario: System restarts and tries to send ALL messages at once!\n');

// Create 1000 pending messages
const messages = [];
for (let i = 0; i < 1000; i++) {
  messages.push({
    id: `msg-${i+1}`,
    content: `Hello message ${i+1}`,
    status: 'pending'
  });
}

let sentCount = 0;
let errorCount = 0;
const startTime = Date.now();

// Simulated server that crashes under load
async function sendToServer() {
  // Server gets overwhelmed with too many concurrent requests
  await sleep(Math.random() * 50);
  
  // High error rate when overwhelmed (50% errors!)
  if (Math.random() < 0.5) {
    errorCount++;
    throw new Error('Server overloaded!');
  }
  
  sentCount++;
  return { success: true };
}

// BAD APPROACH: Send everything at once!
async function sendAllAtOnce() {
  console.log('üí• SENDING ALL 1000 MESSAGES AT ONCE...\n');
  
  const promises = messages.map(async (msg) => {
    try {
      await sendToServer();
      msg.status = 'sent';
    } catch (error) {
      msg.status = 'failed';
    }
  });
  
  // This creates a thundering herd!
  await Promise.all(promises);
  
  const elapsed = (Date.now() - startTime) / 1000;
  
  console.log('='.repeat(60));
  console.log('üí• STORM RESULTS (WITHOUT THROTTLING):');
  console.log('='.repeat(60));
  console.log(`‚ùå Failed: ${errorCount} messages (${(errorCount/1000*100).toFixed(1)}% error rate!)`);
  console.log(`‚úÖ Sent: ${sentCount} messages`);
  console.log(`‚è±Ô∏è  Time: ${elapsed.toFixed(2)} seconds`);
  console.log(`üìà Burst RPS: ${(1000 / elapsed).toFixed(0)} requests/second`);
  console.log('\n‚ö†Ô∏è  PROBLEMS CAUSED:');
  console.log('1. Server overwhelmed - 50% error rate!');
  console.log('2. Massive burst of ' + Math.floor(1000/elapsed) + ' RPS crashed the server');
  console.log('3. Need to retry 500+ failed messages');
  console.log('4. Retries will cause ANOTHER storm!');
  console.log('\nThis is why we need throttling! Run demo.js to see the solution.');
}

sendAllAtOnce().catch(console.error);