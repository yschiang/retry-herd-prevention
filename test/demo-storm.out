============================================================
‚ö†Ô∏è  RETRY STORM DEMO (WITHOUT PROTECTION)
============================================================

Scenario: System restarts and tries to send ALL messages at once!

üí• SENDING ALL 1000 MESSAGES AT ONCE...

============================================================
üí• STORM RESULTS (WITHOUT THROTTLING):
============================================================
‚ùå Failed: 485 messages (48.5% error rate!)
‚úÖ Sent: 515 messages
‚è±Ô∏è  Time: 0.05 seconds
üìà Burst RPS: 20000 requests/second

‚ö†Ô∏è  PROBLEMS CAUSED:
1. Server overwhelmed - 50% error rate!
2. Massive burst of 20000 RPS crashed the server
3. Need to retry 500+ failed messages
4. Retries will cause ANOTHER storm!

This is why we need throttling! Run demo.js to see the solution.
